lib/active_publisher/async/in_memory_adapter/consumer_thread.rb

```diff
@@ -41,10 +40,13 @@ def start_thread
           @thread = ::Thread.new do
             loop do
               # Write "current_messages" so we can requeue should something happen to the consumer.
-              @current_messages << message = queue.pop
+              @current_messages.concat(queue.pop_up_to(20))
 
               begin
-                ::ActivePublisher.publish(message.route, message.payload, message.exchange_name, message.options)
+                # Only open a single connection for each group of messages to an exchange
+                @current_messages.group_by(&:exchange_name).each do |exchange_name, messages|
+                  ::ActivePublisher.publish_all(exchange_name, messages)
```

Comment:
I'm a little bit concerned because this could publish duplicate messages here when we re-concat them upon network error, or spin through them on error.

What if we did something like:

```ruby
def self.publish_all(exchange_name, messages)
  with_exchange(exchange_name) do |exchange|
    loop do
      break if messages.empty?
      message = messages.pop
      fail ActivePublisher::UnknownMessageClassError, "bulk publish messages must be ActivePublisher::Message" unless message.is_a?(ActivePublisher::Message)      
      begin
        exchange.publish(message.payload, publishing_options(message.route, message.options || {}))
      rescue
        messages << message
        raise
      end
    end
  end
end
```
and then do something like..
```ruby
messages_to_retry = []
@current_messages.group_by(&:exchange_name).each do |exchange_name, messages|
  begin
    ::ActivePublisher.publish_all(exchange_name, messages)
  ensure
    # This is mutated by the publish_all (don't love it, but it's an option)
    messages_to_retry.concat(messages)
  end
end

# down where we handle errors...
@current_messages = messages_to_retry
```

This is mostly a sketch, but this way we never drop or publish duplicate messages.